// beau/request.js
const uuid = require('uuid/v4');
const unirest = require('unirest');
const shared = require('./shared');
const RequestList = require('./requestList');
const RequestCache = require('./requestCache');

const httpVerbs = shared.httpVerbs;
const requestRegex = shared.requestRegex;
const replacementRegex = shared.replacementRegex;

class Request {
	constructor(request) {
		let parts = request.request.match(requestRegex);

		this.$id = uuid();
		this.$verb = parts[1];
		this.$dependencies = new Set();
		this.$endpoint = request.HOST + parts[2];

		this.$headers = request.HEADERS;
		this.$payload = request.PAYLOAD;
		this.$params = request.PARAMS;

		if (request.ALIAS) {
			this.$alias = request.ALIAS;
		}

		this.$dependencies = this.findDependencies(request);
	}

	findDependencies(request, set = new Set()) {
		if (typeof request === 'object') {
			Object.keys(request).forEach(key => {
				if (key === 'ALIAS' || key.startsWith('$'))
					return;

				set = this.findDependencies(request[key], set);
			});
		} else {
			let matches = request.match(replacementRegex) || [];
			let deps = matches.map(m => m.split('.')[0]);

			return new Set([...set, ...deps]);
		}

		return set;
	}

	exec(list = new RequestList(), cache = new RequestCache()) {
		let dependencies = [];

		if (this.$dependencies.size > 0) {
			dependencies = Array.from(this.$dependencies).map(dep => {
				return list.execByAlias(dep);
			});
		}

		return Promise.all(dependencies).then(() => {
			let endpoint = cache.parse(this.$endpoint);
			let request = unirest(this.$verb, endpoint);

			request.headers(cache.safely(this.$headers));
			request.query(cache.safely(this.$params));
			request.send(cache.safely(this.$payload));

			return new Promise((resolve, reject) => {
				request.end(res => {
					let results = {
						request: {
							headers: res.request.headers,
							body: res.request.body,
							endpoint: endpoint
						},
						response: {
							status: res.status,
							headers: res.headers,
							body: res.body,
						}
					};


					if (typeof this.$alias !== 'undefined') {
						cache.add(this.$alias, results);
					}

					resolve(results);
				});
			});
		});
	}
}

module.exports = Request;
